File: leetcode/112.path-sum/go/haspathsum.go
package pathsum

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func hasPathSum(root *TreeNode, targetSum int) bool {
	return traversal(root, targetSum, 0)
}

func traversal(root *TreeNode, targetSum int, pathSum int) bool {
	if root == nil {
		return false
	}

	pathSum += root.Val

	if root.Left == nil && root.Right == nil {
		return targetSum == pathSum
	}

	if traversal(root.Left, targetSum, pathSum) {
		return true
	}

	return traversal(root.Right, targetSum, pathSum)
}

File: leetcode/112.path-sum/ts/haspathsum.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

const hasPathSum = (root: TreeNode | null, targetSum: number): boolean => {
    return traversal(root, targetSum, 0)
}

const traversal = (root: TreeNode | null, targetSum: number, pathSum: number): boolean => {
    if (root === null) return false

    pathSum += root.val
    if (root.left === null && root.right === null) return targetSum === pathSum

    if (traversal(root.left, targetSum, pathSum)) return true
    return traversal(root.right, targetSum, pathSum)
}

export default hasPathSum
